### ISSUE TITLE
Reindexing a huge tensor to shuffle it results in data loss at the end of the tensor

### ISSUE THREAD

**Issue by EvenOldridge** [pytorch/pytorch#20888]
## 🐛 Bug

Shuffling a huge tensor by reindexing into it results in all of the values past a certain point being set to 0 when the reindexing is done in GPU memory.

@colesbury  This likely relates to https://github.com/pytorch/pytorch/issues/20562 as the issue is also with reindexing and also shows up at the same point in the tensor.  Hopefully this helps with narrowing the other issue.

I initially thought this was limited to longtensors, but if you use a floattensor with 2x values you run into the exact same issue in the exact same place. 

## To Reproduce

1. Create a huge longtensor (>536870700 values) of random values.  I went with 15Mx45
2. Load it into gpu memory
3. Reindex to shuffle the values

Data past ~536870700 will be all 0.  The data is zero'd mid tensor.

Here's a notebook that reproduces the issue:
https://github.com/EvenOldridge/HugeTensor/blob/master/Huge%20Tensor%20Bug%20-%20Data%20Loss.ipynb

## Expected behavior

Data should not be lost when reindexing

## Environment

- PyTorch Version (e.g., 1.0): 1.01
- OS (e.g., Linux): Linux (Ubuntu 16.04)
- How you installed PyTorch (conda, pip, source): pip
- Python version: 3.6
- CUDA/cuDNN version: 10.0
- GPU models and configuration: TeslaV100 32 Gig (on a DGX-1)

**Comment by ngimel** [comment]
Slight correction - data is not zero'd, data is untouched. 
```
import torch                                                                                                                                 
features = torch.empty(15000000, 45, device="cuda", dtype=torch.long).random_(0, 2**22)                                                                                  
ind=torch.randperm(len(features))                                                                                                            
tmp = torch.ones(15000000,50,device="cuda", dtype=torch.long)
del tmp
features3=features[ind]                                                                                                                      
print( features3[0])
print(features3[14999999] )
```
will print ones for the last row. This https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/IndexKernel.cu#L44 has `int` argument, but I thought idx should be based on the number of elements, and that is less than 2**31. Probably launch_kernel does not handle large tensors.

**Comment by colesbury** [comment]
I'll look at this today. @ngimel do you have any insight into #20562?

**Comment by ngimel** [comment]
No. Isolated repro would help, I think.

**Comment by EvenOldridge** [comment]
Thank you both so much for such a quick fix!  You rock!
